RecoTracker/TkSeedGenerator/plugins/DeepCoreSeedGenerator.cc

```diff
@@ -0,0 +1,583 @@
+
+  for (int x = 0; x < jetDimX; x++) {
+    for (int y = 0; y < jetDimY; y++) {
+      for (int trk = 0; trk < Nover; trk++) {
+        output_combined.second[x][y][trk] =
+            matrix_output_prob(0, x, y, trk, 0);  //outputs.at(1).matrix<double>()(0,x,y,trk);
+
+        for (int p = 0; p < Npar; p++) {
+          output_combined.first[x][y][trk][p] =
+              matrix_output_par(0, x, y, trk, p);  //outputs.at(0).matrix<double>()(0,x,y,trk,p);
+        }
+      }
+    }
+  }
+  return output_combined;
+}
+
+const GeomDet* DeepCoreSeedGenerator::DetectorSelector(int llay,
+                                                       const reco::Candidate& jet,
+                                                       GlobalVector jetDir,
+                                                       const reco::Vertex& jetVertex,
+                                                       const TrackerTopology* const tTopo,
+                                                       const edmNew::DetSetVector<SiPixelCluster>& clusters) {
+  struct trkNumCompare {
+    bool operator()(std::pair<int, const GeomDet*> x, std::pair<int, const GeomDet*> y) const {
+      return x.first > y.first;
+    }
+  };
+  std::set<std::pair<int, const GeomDet*>, trkNumCompare> track4detSet;
+
+  double minDist = 0.0;
+  GeomDet* output = (GeomDet*)nullptr;
+  for (const auto& detset : clusters) {
+    const GeomDet* det = geometry_->idToDet(detset.id());
+    auto aClusterID = detset.id();
+    if (DetId(aClusterID).subdetId() != 1)
+      continue;
+    int lay = tTopo->layer(det->geographicalId());
+    if (lay != llay)
+      continue;
+    std::pair<bool, Basic3DVector<float>> interPair =
+        findIntersection(jetDir, (reco::Candidate::Point)jetVertex.position(), det);
+    if (interPair.first == false)
+      continue;
+    Basic3DVector<float> inter = interPair.second;
+    auto localInter = det->specificSurface().toLocal((GlobalPoint)inter);
+    if ((minDist == 0.0 || std::abs(localInter.x()) < minDist) && std::abs(localInter.y()) < 3.35) {
+      minDist = std::abs(localInter.x());
+      output = (GeomDet*)det;
+    }
+  }  //detset
+  return output;
+}
+std::vector<GlobalVector> DeepCoreSeedGenerator::splittedClusterDirections(
+    const reco::Candidate& jet,
+    const TrackerTopology* const tTopo,
+    const PixelClusterParameterEstimator* pixelCPE,
+    const reco::Vertex& jetVertex,
+    int layer,
+    const edmNew::DetSetVector<SiPixelCluster>& clusters) {
+  std::vector<GlobalVector> clustDirs;
+  for (const auto& detset_int : clusters) {
+    const GeomDet* det_int = geometry_->idToDet(detset_int.id());
+    int lay = tTopo->layer(det_int->geographicalId());
+    if (lay != layer)
+      continue;  //NB: saved bigClusters on all the layers!!
+    for (const auto& aCluster : detset_int) {
+      GlobalPoint clustPos = det_int->surface().toGlobal(
+          pixelCPE->localParametersV(aCluster, (*geometry_->idToDetUnit(detset_int.id())))[0].first);
```

Comment:
perhaps move the idToDetUnit call outside of the loop?

Reply:
ok